source/extensions/access_loggers/wasm/wasm_access_log_impl.h

```diff
@@ -41,8 +43,18 @@ class WasmAccessLog : public AccessLog::Instance {
     tls_slot_ = std::move(tls_slot);
   }
 
+  ~WasmAccessLog() override {
+    if (tls_slot_->currentThreadRegistered() && tls_slot_->get()) {
+      tls_slot_->runOnAllThreads([plugin = plugin_](ThreadLocal::ThreadLocalObjectSharedPtr object)
+                                     -> ThreadLocal::ThreadLocalObjectSharedPtr {
+        object->asType<WasmHandle>().wasm()->startShutdown(plugin);
```

Comment:
Can you merge main to pick up @jmarantz changes on this? We have the typed wrappers now.

Reply:
Done, although it doesn't look like the typed wrappers supports storing `nullptr`.

Reply:
I see you added a layer of indirection through a handle which seems like a waste. We could change runOnAllThreads send in a typed pointer allowing nullptr or absl::optional ref or add a couple of new variants of runOnAllThreads. @PiotrSikora  I'm not sure what the use-case is though for allowing null. Wouldn't you want to always set() your tls container on startup?

@mattklein123 WDYT?

Reply:
It seems like allowing storing nullptr in the slot should be OK? Should we just fix that?

Reply:
We can. What do we prefer:
 * `T*`, null or not
 * `shared_ptr<T>`, null or not
 * `absl::optional<std::reference<T>>` or whatever the syntax for that is :)
 * a pair of new runOnAllThreads methods that pass the object by pointer rather than ref

In a way I'm slightly in favor of the last option because:
 * it involves editing less files
 * the code at the existing call-sites remains very clean

But I also acknowledge that having 6 variants of runOnAllThreads seems less than great.

Reply:
Thinking more, I am now in favor of "`T*`, null or not" and let the caller be responsible for deciding whether a null-check is needed. The advantage is that we can then drop the no-arg variants of runOnAllThreads and just pass in null in those cases. Sorry for the noise.

Reply:
Yeah my preference would be for not having new variants. Maybe I'm missing something but can't we just make it so that shared_ptr that is stored can be null?

Reply:
The shared_ptr can be stored as null, but the call to asType() in the template wrapper does a dynamic cast and ASSERT. That's easy to fix but then we have to pass the thing as something that can be checked for null `T*` or `absl::optional`

Reply:
I see. Yeah I would probably just do optional shared_ptr.

Reply:
Sorry optional reference.

Reply:
https://github.com/envoyproxy/envoy/pull/13883 prototypes that change. Would like to get some quick feedback on the direction prior to filling out comments and adding a unit test for new thin abstraction `struct OptRef`.